// @json-render/ink/dist/catalog.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/catalog.ts
var catalog_exports = {};
__export(catalog_exports, {
  standardActionDefinitions: () => standardActionDefinitions,
  standardComponentDefinitions: () => standardComponentDefinitions
});
module.exports = __toCommonJS(catalog_exports);
var import_zod = require("zod");
var standardComponentDefinitions = {
  // ==========================================================================
  // Layout Components (Ink Primitives)
  // ==========================================================================
  Box: {
    props: import_zod.z.object({
      flexDirection: import_zod.z.enum(["row", "row-reverse", "column", "column-reverse"]).nullable(),
      alignItems: import_zod.z.enum(["flex-start", "center", "flex-end", "stretch"]).nullable(),
      justifyContent: import_zod.z.enum([
        "flex-start",
        "center",
        "flex-end",
        "space-between",
        "space-around",
        "space-evenly"
      ]).nullable(),
      flexGrow: import_zod.z.number().nullable(),
      flexShrink: import_zod.z.number().nullable(),
      flexWrap: import_zod.z.enum(["nowrap", "wrap", "wrap-reverse"]).nullable(),
      width: import_zod.z.union([import_zod.z.number().max(500), import_zod.z.string()]).nullable(),
      height: import_zod.z.union([import_zod.z.number().max(500), import_zod.z.string()]).nullable(),
      minWidth: import_zod.z.union([import_zod.z.number().max(500), import_zod.z.string()]).nullable(),
      minHeight: import_zod.z.union([import_zod.z.number().max(500), import_zod.z.string()]).nullable(),
      padding: import_zod.z.number().nullable(),
      paddingX: import_zod.z.number().nullable(),
      paddingY: import_zod.z.number().nullable(),
      paddingTop: import_zod.z.number().nullable(),
      paddingBottom: import_zod.z.number().nullable(),
      paddingLeft: import_zod.z.number().nullable(),
      paddingRight: import_zod.z.number().nullable(),
      margin: import_zod.z.number().nullable(),
      marginX: import_zod.z.number().nullable(),
      marginY: import_zod.z.number().nullable(),
      marginTop: import_zod.z.number().nullable(),
      marginBottom: import_zod.z.number().nullable(),
      marginLeft: import_zod.z.number().nullable(),
      marginRight: import_zod.z.number().nullable(),
      gap: import_zod.z.number().nullable(),
      columnGap: import_zod.z.number().nullable(),
      rowGap: import_zod.z.number().nullable(),
      borderStyle: import_zod.z.enum([
        "single",
        "double",
        "round",
        "bold",
        "singleDouble",
        "doubleSingle",
        "classic"
      ]).nullable(),
      borderColor: import_zod.z.string().nullable(),
      borderTop: import_zod.z.boolean().nullable(),
      borderBottom: import_zod.z.boolean().nullable(),
      borderLeft: import_zod.z.boolean().nullable(),
      borderRight: import_zod.z.boolean().nullable(),
      borderDimColor: import_zod.z.boolean().nullable(),
      display: import_zod.z.enum(["flex", "none"]).nullable(),
      overflow: import_zod.z.enum(["visible", "hidden"]).nullable(),
      backgroundColor: import_zod.z.string().nullable()
    }),
    slots: ["default"],
    description: "Flexbox layout container (like a terminal <div>). Use for grouping, spacing, borders, and alignment. Default flexDirection is row.",
    example: {
      flexDirection: "column",
      padding: 1,
      gap: 1,
      borderStyle: "round"
    }
  },
  Text: {
    props: import_zod.z.object({
      text: import_zod.z.string(),
      color: import_zod.z.string().nullable(),
      backgroundColor: import_zod.z.string().nullable(),
      bold: import_zod.z.boolean().nullable(),
      italic: import_zod.z.boolean().nullable(),
      underline: import_zod.z.boolean().nullable(),
      strikethrough: import_zod.z.boolean().nullable(),
      dimColor: import_zod.z.boolean().nullable(),
      inverse: import_zod.z.boolean().nullable(),
      wrap: import_zod.z.enum([
        "wrap",
        "truncate",
        "truncate-end",
        "truncate-middle",
        "truncate-start"
      ]).nullable()
    }),
    slots: [],
    description: "Text output with optional styling (color, bold, italic, etc.). Use for all text content in the terminal.",
    example: { text: "Hello, world!", bold: true, color: "green" }
  },
  Newline: {
    props: import_zod.z.object({
      count: import_zod.z.number().nullable()
    }),
    slots: [],
    description: "Inserts one or more blank lines. Must be placed inside a Box with flexDirection column.",
    example: { count: 1 }
  },
  Spacer: {
    props: import_zod.z.object({}),
    slots: [],
    description: "Flexible empty space that expands to fill available room along the main axis. Use between elements to push them apart."
  },
  // ==========================================================================
  // Content Components (Higher-Level)
  // ==========================================================================
  Heading: {
    props: import_zod.z.object({
      text: import_zod.z.string(),
      level: import_zod.z.enum(["h1", "h2", "h3", "h4"]).nullable(),
      color: import_zod.z.string().nullable()
    }),
    slots: [],
    description: "Section heading. h1 is bold + underlined, h2 is bold, h3 is bold + dimmed, h4 is dimmed.",
    example: { text: "Dashboard", level: "h1" }
  },
  Divider: {
    props: import_zod.z.object({
      character: import_zod.z.string().nullable(),
      color: import_zod.z.string().nullable(),
      dimColor: import_zod.z.boolean().nullable(),
      title: import_zod.z.string().nullable(),
      width: import_zod.z.number().max(500).nullable()
    }),
    slots: [],
    description: "Horizontal separator line. Default width is 40 characters. Optionally includes a centered title.",
    example: { title: "Section", color: "gray", width: 40 }
  },
  Badge: {
    props: import_zod.z.object({
      label: import_zod.z.string(),
      variant: import_zod.z.enum(["default", "info", "success", "warning", "error"]).nullable()
    }),
    slots: [],
    description: "Small colored inline label for status, counts, and categories.",
    example: { label: "ACTIVE", variant: "success" }
  },
  Spinner: {
    props: import_zod.z.object({
      label: import_zod.z.string().nullable(),
      color: import_zod.z.string().nullable()
    }),
    slots: [],
    description: "Animated loading spinner with optional label text. Uses braille animation characters.",
    example: { label: "Loading...", color: "cyan" }
  },
  ProgressBar: {
    props: import_zod.z.object({
      progress: import_zod.z.number(),
      width: import_zod.z.number().max(500).nullable(),
      color: import_zod.z.string().nullable(),
      label: import_zod.z.string().nullable()
    }),
    slots: [],
    description: "Horizontal progress bar. Set progress from 0 to 1. Default width is 30 characters.",
    example: { progress: 0.65, width: 30, color: "green", label: "Uploading" }
  },
  Sparkline: {
    props: import_zod.z.object({
      data: import_zod.z.array(import_zod.z.number()),
      width: import_zod.z.number().max(500).nullable(),
      color: import_zod.z.string().nullable(),
      label: import_zod.z.string().nullable(),
      min: import_zod.z.number().nullable(),
      max: import_zod.z.number().nullable()
    }),
    slots: [],
    description: "Inline sparkline chart using Unicode block characters (\u2581\u2582\u2583\u2584\u2585\u2586\u2587\u2588). Pass an array of numbers to visualize trends compactly. Set min/max to fix the scale across multiple sparklines.",
    example: {
      data: [3, 7, 2, 9, 4, 8, 1, 6, 5],
      color: "cyan",
      label: "CPU"
    }
  },
  BarChart: {
    props: import_zod.z.object({
      data: import_zod.z.array(
        import_zod.z.object({
          label: import_zod.z.string(),
          value: import_zod.z.number(),
          color: import_zod.z.string().nullable()
        })
      ),
      width: import_zod.z.number().max(500).nullable(),
      showValues: import_zod.z.boolean().nullable(),
      showPercentage: import_zod.z.boolean().nullable()
    }),
    slots: [],
    description: "Horizontal bar chart. Each item has a label, numeric value, and optional color. Set showValues to display raw numbers, showPercentage to show % of total. Default bar width is 30.",
    example: {
      data: [
        { label: "TypeScript", value: 65, color: "blue" },
        { label: "Python", value: 20, color: "yellow" },
        { label: "Rust", value: 15, color: "red" }
      ],
      showPercentage: true
    }
  },
  Table: {
    props: import_zod.z.object({
      columns: import_zod.z.array(
        import_zod.z.object({
          header: import_zod.z.string(),
          key: import_zod.z.string(),
          width: import_zod.z.number().max(200).nullable(),
          align: import_zod.z.enum(["left", "center", "right"]).nullable()
        })
      ),
      rows: import_zod.z.array(import_zod.z.record(import_zod.z.string(), import_zod.z.string())),
      borderStyle: import_zod.z.enum(["single", "double", "round", "bold", "classic"]).nullable(),
      backgroundColor: import_zod.z.string().nullable(),
      headerColor: import_zod.z.string().nullable()
    }),
    slots: [],
    description: "Tabular data display with headers and rows. Each row is a record mapping column keys to string values. Set both borderStyle and backgroundColor together so borders share the same shading.",
    example: {
      columns: [
        { header: "Name", key: "name", width: 20 },
        { header: "Status", key: "status", width: 10 }
      ],
      rows: [
        { name: "api-server", status: "running" },
        { name: "worker", status: "stopped" }
      ],
      headerColor: "cyan"
    }
  },
  List: {
    props: import_zod.z.object({
      items: import_zod.z.array(import_zod.z.string()),
      ordered: import_zod.z.boolean().nullable(),
      bulletChar: import_zod.z.string().nullable(),
      spacing: import_zod.z.number().nullable()
    }),
    slots: [],
    description: "Bulleted or numbered list. Each item is a string. Use for simple enumerations.",
    example: {
      items: ["Install dependencies", "Run tests", "Deploy"],
      ordered: true
    }
  },
  ListItem: {
    props: import_zod.z.object({
      title: import_zod.z.string(),
      subtitle: import_zod.z.string().nullable(),
      leading: import_zod.z.string().nullable(),
      trailing: import_zod.z.string().nullable()
    }),
    slots: [],
    description: "Structured list row with title, optional subtitle, and leading/trailing text. Use with repeat for dynamic lists.",
    example: {
      title: "package.json",
      subtitle: "Modified 2 hours ago",
      leading: "*",
      trailing: "2.1 KB"
    }
  },
  Card: {
    props: import_zod.z.object({
      title: import_zod.z.string().nullable(),
      backgroundColor: import_zod.z.string().nullable(),
      padding: import_zod.z.number().nullable()
    }),
    slots: ["default"],
    description: "Shaded container with optional title. Renders as a filled background area for grouping related content. Default background is a subtle dark shade.",
    example: { title: "Details", padding: 1 }
  },
  KeyValue: {
    props: import_zod.z.object({
      label: import_zod.z.string(),
      value: import_zod.z.union([import_zod.z.string(), import_zod.z.number(), import_zod.z.array(import_zod.z.string())]),
      labelColor: import_zod.z.string().nullable(),
      separator: import_zod.z.string().nullable()
    }),
    slots: [],
    description: "Key-value pair display. Renders label and value on the same line. Value can be a string, number, or array of strings (joined with commas). Default separator is a colon.",
    example: { label: "Status", value: "Running", labelColor: "cyan" }
  },
  Link: {
    props: import_zod.z.object({
      url: import_zod.z.string(),
      label: import_zod.z.string().nullable(),
      color: import_zod.z.string().nullable()
    }),
    slots: [],
    description: 'Renders a URL as underlined text. If label is provided, shows "label (url)". Most terminals make URLs clickable automatically.',
    example: { url: "https://github.com/vercel/next.js", label: "Next.js" }
  },
  StatusLine: {
    props: import_zod.z.object({
      text: import_zod.z.string(),
      status: import_zod.z.enum(["info", "success", "warning", "error"]).nullable(),
      icon: import_zod.z.string().nullable()
    }),
    slots: [],
    description: "Status message with colored icon. Default icons: info=\u2139, success=\u2714, warning=\u26A0, error=\u2716.",
    example: { text: "Build completed successfully", status: "success" }
  },
  Metric: {
    props: import_zod.z.object({
      label: import_zod.z.string(),
      value: import_zod.z.string(),
      detail: import_zod.z.string().nullable(),
      trend: import_zod.z.enum(["up", "down", "neutral"]).nullable()
    }),
    slots: [],
    description: "Key metric display with prominent value and optional trend indicator. Use for important numbers that deserve visual emphasis (price, temperature, stars, market cap).",
    example: {
      label: "Price",
      value: "$70,686",
      detail: "24h change",
      trend: "up"
    }
  },
  Callout: {
    props: import_zod.z.object({
      type: import_zod.z.enum(["info", "tip", "warning", "important"]).nullable(),
      title: import_zod.z.string().nullable(),
      content: import_zod.z.string()
    }),
    slots: [],
    description: "Highlighted callout block with colored left border. Use for key takeaways, tips, warnings, or important notes that should stand out from surrounding content.",
    example: {
      type: "tip",
      title: "Key Takeaway",
      content: "Revenue peaked in FY2022 driven by iPhone 13/14 upgrade cycles."
    }
  },
  Timeline: {
    props: import_zod.z.object({
      items: import_zod.z.array(
        import_zod.z.object({
          title: import_zod.z.string(),
          description: import_zod.z.string().nullable(),
          date: import_zod.z.string().nullable(),
          status: import_zod.z.enum(["completed", "current", "upcoming"]).nullable()
        })
      )
    }),
    slots: [],
    description: "Vertical timeline showing ordered events, steps, or milestones. Each item has a status-colored dot, title, optional date, and optional description.",
    example: {
      items: [
        {
          title: "Project Started",
          description: "Initial commit and setup",
          date: "Jan 2024",
          status: "completed"
        },
        {
          title: "Beta Release",
          description: "Public beta launched",
          date: "Mar 2024",
          status: "current"
        },
        { title: "v1.0", date: "Q2 2024", status: "upcoming" }
      ]
    }
  },
  // ==========================================================================
  // Interactive Components
  // ==========================================================================
  TextInput: {
    props: import_zod.z.object({
      placeholder: import_zod.z.string().nullable(),
      value: import_zod.z.string().nullable(),
      label: import_zod.z.string().nullable(),
      mask: import_zod.z.string().nullable()
    }),
    events: ["submit", "change"],
    slots: [],
    description: "Text input field for terminal. Use $bindState to bind to the state model for two-way binding. Press Enter to submit. Set mask to '*' for password fields.",
    example: { placeholder: "Type here...", label: "Name" }
  },
  Select: {
    props: import_zod.z.object({
      options: import_zod.z.array(
        import_zod.z.object({
          label: import_zod.z.string(),
          value: import_zod.z.string()
        })
      ),
      value: import_zod.z.string().nullable(),
      label: import_zod.z.string().nullable()
    }),
    events: ["change"],
    slots: [],
    description: "Selection menu navigated with arrow keys. Use $bindState on value to bind the selected value to state. Press Enter to confirm selection.",
    example: {
      options: [
        { label: "Development", value: "dev" },
        { label: "Staging", value: "staging" },
        { label: "Production", value: "prod" }
      ],
      label: "Environment"
    }
  },
  MultiSelect: {
    props: import_zod.z.object({
      options: import_zod.z.array(
        import_zod.z.object({
          label: import_zod.z.string(),
          value: import_zod.z.string()
        })
      ),
      value: import_zod.z.array(import_zod.z.string()).nullable(),
      label: import_zod.z.string().nullable(),
      min: import_zod.z.number().nullable(),
      max: import_zod.z.number().nullable()
    }),
    events: ["change", "submit"],
    slots: [],
    description: "Multi-selection menu. Navigate with arrow keys, toggle with space, confirm with enter. Use $bindState on value to bind the selected values array to state. Set min/max to constrain selection count.",
    example: {
      options: [
        { label: "TypeScript", value: "ts" },
        { label: "Python", value: "py" },
        { label: "Rust", value: "rs" },
        { label: "Go", value: "go" }
      ],
      label: "Languages"
    }
  },
  ConfirmInput: {
    props: import_zod.z.object({
      message: import_zod.z.string().nullable(),
      defaultValue: import_zod.z.boolean().nullable(),
      yesLabel: import_zod.z.string().nullable(),
      noLabel: import_zod.z.string().nullable()
    }),
    events: ["confirm", "deny"],
    slots: [],
    description: "Yes/No confirmation prompt. Press Y to confirm, N to deny. Use for destructive or irreversible actions.",
    example: {
      message: "Delete all files?"
    }
  },
  Tabs: {
    props: import_zod.z.object({
      tabs: import_zod.z.array(
        import_zod.z.object({
          label: import_zod.z.string(),
          value: import_zod.z.string(),
          icon: import_zod.z.string().nullable()
        })
      ),
      value: import_zod.z.string().nullable(),
      color: import_zod.z.string().nullable()
    }),
    events: ["change"],
    slots: ["default"],
    description: "Tab bar navigation. Navigate with left/right arrow keys. Use $bindState on value to bind the active tab to state. Place child content inside and use visible conditions on children to show content for the active tab.",
    example: {
      tabs: [
        { label: "Overview", value: "overview" },
        { label: "Logs", value: "logs" },
        { label: "Settings", value: "settings" }
      ]
    }
  },
  Markdown: {
    props: import_zod.z.object({
      text: import_zod.z.string()
    }),
    slots: [],
    description: "Renders markdown-formatted text with proper terminal styling. Supports headings (#), **bold**, *italic*, `inline code`, ~~strikethrough~~, fenced code blocks, lists (ordered and unordered), blockquotes (>), and horizontal rules (---).",
    example: {
      text: "## Overview\n\nThis is **bold** and *italic* text with `inline code`.\n\n- First item\n- Second item\n\n> A blockquote"
    }
  }
};
var standardActionDefinitions = {
  setState: {
    params: import_zod.z.object({
      statePath: import_zod.z.string(),
      value: import_zod.z.unknown()
    }),
    description: "Update a value in the state model at the given statePath."
  },
  pushState: {
    params: import_zod.z.object({
      statePath: import_zod.z.string(),
      value: import_zod.z.unknown(),
      clearStatePath: import_zod.z.string().optional()
    }),
    description: 'Append an item to an array in the state model. The value can contain { $state: "/statePath" } references and "$id" for auto IDs. Use clearStatePath to reset another path after pushing.'
  },
  removeState: {
    params: import_zod.z.object({
      statePath: import_zod.z.string(),
      index: import_zod.z.number()
    }),
    description: "Remove an item from an array in the state model at the given index."
  },
  exit: {
    params: import_zod.z.object({
      code: import_zod.z.number().optional()
    }),
    description: "Exit the terminal application. Optional exit code (default 0)."
  },
  log: {
    params: import_zod.z.object({
      message: import_zod.z.string()
    }),
    description: "Write a message to stdout outside the Ink render. Useful for persistent output that should remain visible after the UI updates."
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  standardActionDefinitions,
  standardComponentDefinitions
});
