// @json-render/ink/dist/index.mjs
import {
  standardActionDefinitions,
  standardComponentDefinitions
} from "./chunk-JOOY7VDB.mjs";
import {
  schema
} from "./chunk-WESCBWNK.mjs";

// ../react-state/dist/index.mjs
import {
  createContext,
  useContext,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useSyncExternalStore
} from "react";
import {
  getByPath,
  createStateStore
} from "@json-render/core";
import { flattenToPointers } from "@json-render/core/store-utils";
import { jsx } from "react/jsx-runtime";
var StateContext = createContext(null);
function computeInitialFlat(isControlled, initialState) {
  if (isControlled) return null;
  if (Object.keys(initialState).length === 0) return {};
  return flattenToPointers(initialState);
}
function StateProvider({
  store: externalStore,
  initialState = {},
  onStateChange,
  children
}) {
  const internalStoreRef = useRef(void 0);
  if (!externalStore && !internalStoreRef.current) {
    internalStoreRef.current = createStateStore(initialState);
  }
  const store = externalStore ?? internalStoreRef.current;
  const storeRef = useRef(store);
  storeRef.current = store;
  const isControlledRef = useRef(!!externalStore);
  isControlledRef.current = !!externalStore;
  const initialModeRef = useRef(externalStore ? "controlled" : "uncontrolled");
  const modeWarnedRef = useRef(false);
  if (process.env.NODE_ENV !== "production") {
    const currentMode = externalStore ? "controlled" : "uncontrolled";
    if (currentMode !== initialModeRef.current && !modeWarnedRef.current) {
      modeWarnedRef.current = true;
      console.warn(
        `StateProvider: switching from ${initialModeRef.current} to ${currentMode} mode is not supported.`
      );
    }
  }
  const prevInitialStateRef = useRef(initialState);
  const prevFlatRef = useRef(
    computeInitialFlat(!!externalStore, initialState)
  );
  useEffect(() => {
    if (externalStore) return;
    if (initialState === prevInitialStateRef.current) return;
    prevInitialStateRef.current = initialState;
    const nextFlat = initialState && Object.keys(initialState).length > 0 ? flattenToPointers(initialState) : {};
    const prevFlat = prevFlatRef.current ?? {};
    const allKeys = /* @__PURE__ */ new Set([
      ...Object.keys(prevFlat),
      ...Object.keys(nextFlat)
    ]);
    const updates = {};
    for (const key of allKeys) {
      if (prevFlat[key] !== nextFlat[key]) {
        updates[key] = key in nextFlat ? nextFlat[key] : void 0;
      }
    }
    prevFlatRef.current = nextFlat;
    if (Object.keys(updates).length > 0) {
      store.update(updates);
    }
  }, [externalStore, initialState, store]);
  const state = useSyncExternalStore(
    store.subscribe,
    store.getSnapshot,
    store.getServerSnapshot ?? store.getSnapshot
  );
  const onStateChangeRef = useRef(onStateChange);
  onStateChangeRef.current = onStateChange;
  const set = useCallback((path, value2) => {
    const s = storeRef.current;
    const prev = s.getSnapshot();
    s.set(path, value2);
    if (!isControlledRef.current && s.getSnapshot() !== prev) {
      onStateChangeRef.current?.([{ path, value: value2 }]);
    }
  }, []);
  const update = useCallback((updates) => {
    const s = storeRef.current;
    const prev = s.getSnapshot();
    s.update(updates);
    if (!isControlledRef.current && s.getSnapshot() !== prev) {
      const changes = [];
      for (const [path, value2] of Object.entries(updates)) {
        if (getByPath(prev, path) !== value2) {
          changes.push({ path, value: value2 });
        }
      }
      if (changes.length > 0) {
        onStateChangeRef.current?.(changes);
      }
    }
  }, []);
  const get = useCallback((path) => storeRef.current.get(path), []);
  const getSnapshot = useCallback(() => storeRef.current.getSnapshot(), []);
  const value = useMemo(
    () => ({ state, get, set, update, getSnapshot }),
    [state, get, set, update, getSnapshot]
  );
  return /* @__PURE__ */ jsx(StateContext.Provider, { value, children });
}
function useStateStore() {
  const ctx = useContext(StateContext);
  if (!ctx) {
    throw new Error("useStateStore must be used within a StateProvider");
  }
  return ctx;
}
function useStateValue(path) {
  const { state } = useStateStore();
  return getByPath(state, path);
}
function useStateBinding(path) {
  const { state, set } = useStateStore();
  const value = getByPath(state, path);
  const setValue = useCallback(
    (newValue) => set(path, newValue),
    [path, set]
  );
  return [value, setValue];
}

// src/contexts/visibility.tsx
import {
  createContext as createContext2,
  useContext as useContext2,
  useMemo as useMemo2
} from "react";
import {
  evaluateVisibility
} from "@json-render/core";
import { jsx as jsx2 } from "react/jsx-runtime";
var VisibilityContext = createContext2(null);
function VisibilityProvider({ children }) {
  const { state } = useStateStore();
  const ctx = useMemo2(
    () => ({
      stateModel: state
    }),
    [state]
  );
  const isVisible = useMemo2(
    () => (condition) => evaluateVisibility(condition, ctx),
    [ctx]
  );
  const value = useMemo2(
    () => ({ isVisible, ctx }),
    [isVisible, ctx]
  );
  return /* @__PURE__ */ jsx2(VisibilityContext.Provider, { value, children });
}
function useVisibility() {
  const ctx = useContext2(VisibilityContext);
  if (!ctx) {
    throw new Error("useVisibility must be used within a VisibilityProvider");
  }
  return ctx;
}
function useIsVisible(condition) {
  const { isVisible } = useVisibility();
  return isVisible(condition);
}

// src/contexts/actions.tsx
import {
  createContext as createContext4,
  useContext as useContext4,
  useState as useState2,
  useCallback as useCallback3,
  useMemo as useMemo4,
  useRef as useRef2
} from "react";
import { Box, Text, useInput as useInput2 } from "ink";
import {
  resolveAction,
  executeAction
} from "@json-render/core";

// src/contexts/focus.tsx
import {
  createContext as createContext3,
  useCallback as useCallback2,
  useContext as useContext3,
  useEffect as useEffect2,
  useId,
  useMemo as useMemo3,
  useReducer,
  useState
} from "react";
import { useInput } from "ink";
import { jsx as jsx3 } from "react/jsx-runtime";
var FocusContext = createContext3(null);
function focusReducer(state, action) {
  switch (action.type) {
    case "register": {
      if (state.ids.includes(action.id)) return state;
      const ids = [...state.ids, action.id];
      const focusedId = state.focusedId ?? action.id;
      return { ids, focusedId };
    }
    case "unregister": {
      const ids = state.ids.filter((x) => x !== action.id);
      let focusedId = state.focusedId;
      if (focusedId === action.id) {
        focusedId = ids[0] ?? null;
      }
      return { ids, focusedId };
    }
    case "cycle": {
      const { ids, focusedId } = state;
      if (ids.length === 0) return { ...state, focusedId: null };
      const idx = focusedId ? ids.indexOf(focusedId) : -1;
      const next = action.direction === "next" ? ids[(idx + 1) % ids.length] : ids[(idx - 1 + ids.length) % ids.length];
      return { ...state, focusedId: next };
    }
  }
}
function FocusProvider({ children }) {
  const [state, dispatch] = useReducer(focusReducer, {
    ids: [],
    focusedId: null
  });
  const [disableCount, setDisableCount] = useState(0);
  const disabled = disableCount > 0;
  const setDisabled = useCallback2((value2) => {
    setDisableCount((prev) => value2 ? prev + 1 : Math.max(0, prev - 1));
  }, []);
  const register = useCallback2(
    (id) => dispatch({ type: "register", id }),
    []
  );
  const unregister = useCallback2(
    (id) => dispatch({ type: "unregister", id }),
    []
  );
  const isFocused = useCallback2(
    (id) => !disabled && state.focusedId === id,
    [state.focusedId, disabled]
  );
  useInput((_input, key) => {
    if (disabled) return;
    if (!key.tab) return;
    dispatch({ type: "cycle", direction: key.shift ? "prev" : "next" });
  });
  const value = useMemo3(
    () => ({
      register,
      unregister,
      focusedId: state.focusedId,
      isFocused,
      setDisabled
    }),
    [register, unregister, state.focusedId, isFocused, setDisabled]
  );
  return /* @__PURE__ */ jsx3(FocusContext.Provider, { value, children });
}
function useFocus() {
  const ctx = useContext3(FocusContext);
  const id = useId();
  useEffect2(() => {
    ctx?.register(id);
    return () => {
      ctx?.unregister(id);
    };
  }, [ctx, id]);
  const isActive = ctx?.isFocused(id) ?? false;
  return { isActive, id };
}
function useFocusDisable(disabled) {
  const ctx = useContext3(FocusContext);
  useEffect2(() => {
    if (!disabled) return;
    ctx?.setDisabled(true);
    return () => {
      ctx?.setDisabled(false);
    };
  }, [ctx, disabled]);
}

// src/contexts/actions.tsx
import { jsx as jsx4, jsxs } from "react/jsx-runtime";
function generateUniqueId() {
  if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
    return crypto.randomUUID();
  }
  return `${Date.now()}-${Math.random().toString(36).slice(2)}`;
}
var MAX_RESOLVE_DEPTH = 10;
function deepResolveValue(value, get, depth = 0) {
  if (depth > MAX_RESOLVE_DEPTH) return value;
  if (value === null || value === void 0) return value;
  if (value === "$id") {
    return generateUniqueId();
  }
  if (typeof value === "object" && !Array.isArray(value)) {
    const obj = value;
    const keys = Object.keys(obj);
    if (keys.length === 1 && typeof obj.$state === "string") {
      return get(obj.$state);
    }
    if (keys.length === 1 && "$id" in obj) {
      return generateUniqueId();
    }
    const resolved = {};
    for (const [key, val] of Object.entries(obj)) {
      resolved[key] = deepResolveValue(val, get, depth + 1);
    }
    return resolved;
  }
  if (Array.isArray(value)) {
    return value.map((item) => deepResolveValue(item, get, depth + 1));
  }
  return value;
}
var ActionContext = createContext4(null);
function ActionProvider({
  handlers: initialHandlers = {},
  navigate,
  children
}) {
  const { get, set, getSnapshot } = useStateStore();
  const initialHandlersRef = useRef2(initialHandlers);
  initialHandlersRef.current = initialHandlers;
  const navigateRef = useRef2(navigate);
  navigateRef.current = navigate;
  const [loadingActions, setLoadingActions] = useState2(
    /* @__PURE__ */ new Map()
  );
  const [pendingConfirmation, setPendingConfirmation] = useState2(null);
  const pendingRef = useRef2(null);
  const lookupHandler = useCallback3(
    (name) => initialHandlersRef.current[name],
    []
  );
  const execute = useCallback3(
    async (binding) => {
      const resolved = resolveAction(binding, getSnapshot());
      const customHandler = lookupHandler(resolved.action);
      if (resolved.action === "setState" && !customHandler && resolved.params) {
        const statePath = resolved.params.statePath;
        const value2 = resolved.params.value;
        if (statePath) {
          set(statePath, value2);
        }
        return;
      }
      if (resolved.action === "pushState" && !customHandler && resolved.params) {
        const statePath = resolved.params.statePath;
        const rawValue = resolved.params.value;
        if (statePath) {
          const resolvedValue = deepResolveValue(rawValue, get);
          const raw = get(statePath);
          const arr = Array.isArray(raw) ? raw : [];
          set(statePath, [...arr, resolvedValue]);
          const clearStatePath = resolved.params.clearStatePath;
          if (clearStatePath) {
            set(clearStatePath, "");
          }
        }
        return;
      }
      if (resolved.action === "removeState" && !customHandler && resolved.params) {
        const statePath = resolved.params.statePath;
        const index = resolved.params.index;
        if (statePath !== void 0 && index !== void 0) {
          const raw = get(statePath);
          if (!Array.isArray(raw)) return;
          set(
            statePath,
            raw.filter((_, i) => i !== index)
          );
        }
        return;
      }
      if (resolved.action === "log" && !customHandler) {
        const message = resolved.params?.message ?? resolved.params?.value ?? "";
        console.log("[json-render]", message);
        return;
      }
      if (resolved.action === "exit" && !customHandler) {
        return;
      }
      const handler = customHandler;
      if (!handler) {
        console.warn(
          `[json-render] No handler registered for action: ${resolved.action}`
        );
        return;
      }
      if (resolved.confirm) {
        pendingRef.current?.reject();
        const confirmed = await new Promise((res) => {
          const entry = {
            action: resolved,
            handler,
            resolve: () => {
              pendingRef.current = null;
              setPendingConfirmation(null);
              res(true);
            },
            reject: () => {
              pendingRef.current = null;
              setPendingConfirmation(null);
              res(false);
            }
          };
          pendingRef.current = entry;
          setPendingConfirmation(entry);
        });
        if (!confirmed) return;
      }
      const actionName = resolved.action;
      setLoadingActions((prev) => {
        const next = new Map(prev);
        next.set(actionName, (next.get(actionName) ?? 0) + 1);
        return next;
      });
      try {
        await executeAction({
          action: resolved,
          handler,
          setState: set,
          navigate: navigateRef.current,
          executeAction: async (name) => {
            const subBinding = { action: name };
            await execute(subBinding);
          }
        });
      } finally {
        setLoadingActions((prev) => {
          const next = new Map(prev);
          const count = (next.get(actionName) ?? 1) - 1;
          if (count <= 0) {
            next.delete(actionName);
          } else {
            next.set(actionName, count);
          }
          return next;
        });
      }
    },
    [lookupHandler, get, set, getSnapshot]
  );
  const confirm = useCallback3(() => {
    pendingRef.current?.resolve();
  }, []);
  const cancel = useCallback3(() => {
    pendingRef.current?.reject();
  }, []);
  const value = useMemo4(
    () => ({
      handlers: initialHandlers,
      loadingActions,
      pendingConfirmation,
      execute,
      confirm,
      cancel
    }),
    [
      initialHandlers,
      loadingActions,
      pendingConfirmation,
      execute,
      confirm,
      cancel
    ]
  );
  return /* @__PURE__ */ jsx4(ActionContext.Provider, { value, children });
}
function useActions() {
  const ctx = useContext4(ActionContext);
  if (!ctx) {
    throw new Error("useActions must be used within an ActionProvider");
  }
  return ctx;
}
function useAction(binding) {
  const { execute, loadingActions } = useActions();
  const executeAction2 = useCallback3(() => execute(binding), [execute, binding]);
  const isLoading = (loadingActions.get(binding.action) ?? 0) > 0;
  return { execute: executeAction2, isLoading };
}
function ConfirmDialog({
  confirm,
  onConfirm,
  onCancel
}) {
  const isDanger = confirm.variant === "danger";
  useFocusDisable(true);
  useInput2((input, key) => {
    if (input.toLowerCase() === "y") {
      onConfirm();
    } else if (input.toLowerCase() === "n" || key.escape) {
      onCancel();
    }
  });
  return /* @__PURE__ */ jsxs(
    Box,
    {
      flexDirection: "column",
      borderStyle: "round",
      borderColor: isDanger ? "red" : "blue",
      paddingX: 2,
      paddingY: 1,
      children: [
        /* @__PURE__ */ jsx4(Text, { bold: true, children: confirm.title }),
        /* @__PURE__ */ jsx4(Text, { dimColor: true, children: confirm.message }),
        /* @__PURE__ */ jsxs(Box, { marginTop: 1, gap: 2, children: [
          /* @__PURE__ */ jsxs(Text, { children: [
            /* @__PURE__ */ jsx4(Text, { color: isDanger ? "red" : "green", bold: true, children: "[Y]" }),
            " ",
            confirm.confirmLabel ?? "Confirm"
          ] }),
          /* @__PURE__ */ jsxs(Text, { children: [
            /* @__PURE__ */ jsx4(Text, { bold: true, children: "[N]" }),
            " ",
            confirm.cancelLabel ?? "Cancel"
          ] })
        ] })
      ]
    }
  );
}

// src/contexts/validation.tsx
import {
  createContext as createContext5,
  useContext as useContext5,
  useState as useState3,
  useCallback as useCallback4,
  useEffect as useEffect3,
  useMemo as useMemo5
} from "react";
import {
  runValidation
} from "@json-render/core";
import { jsx as jsx5 } from "react/jsx-runtime";
var ValidationContext = createContext5(null);
function dynamicArgsEqual(a, b) {
  if (a === b) return true;
  if (!a || !b) return false;
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);
  if (keysA.length !== keysB.length) return false;
  for (const key of keysA) {
    const va = a[key];
    const vb = b[key];
    if (va === vb) continue;
    if (typeof va === "object" && va !== null && typeof vb === "object" && vb !== null) {
      const sa = va.$state;
      const sb = vb.$state;
      if (typeof sa === "string" && sa === sb) continue;
    }
    return false;
  }
  return true;
}
function validationConfigEqual(a, b) {
  if (a === b) return true;
  if (a.validateOn !== b.validateOn) return false;
  const ac = a.checks ?? [];
  const bc = b.checks ?? [];
  if (ac.length !== bc.length) return false;
  for (let i = 0; i < ac.length; i++) {
    const ca = ac[i];
    const cb = bc[i];
    if (ca.type !== cb.type) return false;
    if (ca.message !== cb.message) return false;
    if (!dynamicArgsEqual(ca.args, cb.args)) return false;
  }
  return true;
}
function ValidationProvider({
  customFunctions = {},
  children
}) {
  const { getSnapshot } = useStateStore();
  const [fieldStates, setFieldStates] = useState3({});
  const [fieldConfigs, setFieldConfigs] = useState3({});
  const registerField = useCallback4(
    (path, config) => {
      setFieldConfigs((prev) => {
        const existing = prev[path];
        if (existing && validationConfigEqual(existing, config)) {
          return prev;
        }
        return { ...prev, [path]: config };
      });
    },
    []
  );
  const unregisterField = useCallback4((path) => {
    setFieldConfigs((prev) => {
      const { [path]: _, ...rest } = prev;
      return rest;
    });
    setFieldStates((prev) => {
      const { [path]: _, ...rest } = prev;
      return rest;
    });
  }, []);
  const validate = useCallback4(
    (path, config) => {
      const currentState = getSnapshot();
      const segments = path.split("/").filter(Boolean);
      let value2 = currentState;
      for (const seg of segments) {
        if (value2 != null && typeof value2 === "object") {
          value2 = value2[seg];
        } else {
          value2 = void 0;
          break;
        }
      }
      const result = runValidation(config, {
        value: value2,
        stateModel: currentState,
        customFunctions
      });
      setFieldStates((prev) => ({
        ...prev,
        [path]: {
          touched: prev[path]?.touched ?? true,
          validated: true,
          result
        }
      }));
      return result;
    },
    [getSnapshot, customFunctions]
  );
  const touch = useCallback4((path) => {
    setFieldStates((prev) => ({
      ...prev,
      [path]: {
        ...prev[path],
        touched: true,
        validated: prev[path]?.validated ?? false,
        result: prev[path]?.result ?? null
      }
    }));
  }, []);
  const clear = useCallback4((path) => {
    setFieldStates((prev) => {
      const { [path]: _, ...rest } = prev;
      return rest;
    });
  }, []);
  const validateAll = useCallback4(() => {
    let allValid = true;
    for (const [path, config] of Object.entries(fieldConfigs)) {
      const result = validate(path, config);
      if (!result.valid) {
        allValid = false;
      }
    }
    return allValid;
  }, [fieldConfigs, validate]);
  const value = useMemo5(
    () => ({
      customFunctions,
      fieldStates,
      validate,
      touch,
      clear,
      validateAll,
      registerField,
      unregisterField
    }),
    [
      customFunctions,
      fieldStates,
      validate,
      touch,
      clear,
      validateAll,
      registerField,
      unregisterField
    ]
  );
  return /* @__PURE__ */ jsx5(ValidationContext.Provider, { value, children });
}
function useValidation() {
  const ctx = useContext5(ValidationContext);
  if (!ctx) {
    throw new Error("useValidation must be used within a ValidationProvider");
  }
  return ctx;
}
function useOptionalValidation() {
  return useContext5(ValidationContext);
}
function useFieldValidation(path, config) {
  const {
    fieldStates,
    validate: validateField,
    touch: touchField,
    clear: clearField,
    registerField,
    unregisterField
  } = useValidation();
  useEffect3(() => {
    if (config) {
      registerField(path, config);
    }
  }, [path, config, registerField]);
  useEffect3(() => {
    return () => {
      unregisterField(path);
    };
  }, [path]);
  const state = fieldStates[path] ?? {
    touched: false,
    validated: false,
    result: null
  };
  const validate = useCallback4(
    () => validateField(path, config ?? { checks: [] }),
    [path, config, validateField]
  );
  const touch = useCallback4(() => touchField(path), [path, touchField]);
  const clear = useCallback4(() => clearField(path), [path, clearField]);
  return {
    state,
    validate,
    touch,
    clear,
    errors: state.result?.errors ?? [],
    isValid: state.result?.valid ?? true
  };
}

// src/contexts/repeat-scope.tsx
import { createContext as createContext6, useContext as useContext6, useMemo as useMemo6 } from "react";
import { jsx as jsx6 } from "react/jsx-runtime";
var RepeatScopeContext = createContext6(null);
function RepeatScopeProvider({
  item,
  index,
  basePath,
  children
}) {
  const value = useMemo6(
    () => ({ item, index, basePath }),
    [item, index, basePath]
  );
  return /* @__PURE__ */ jsx6(RepeatScopeContext.Provider, { value, children });
}
function useRepeatScope() {
  return useContext6(RepeatScopeContext);
}

// src/index.ts
import { createStateStore as createStateStore2 } from "@json-render/core";

// src/renderer.tsx
import React3, {
  useCallback as useCallback6,
  useMemo as useMemo8
} from "react";
import { Text as Text3 } from "ink";
import {
  resolveElementProps,
  resolveBindings,
  resolveActionParam,
  evaluateVisibility as evaluateVisibility2,
  getByPath as getByPath2
} from "@json-render/core";

// src/components/standard.tsx
import { useState as useState5, useEffect as useEffect5, useMemo as useMemo7 } from "react";
import {
  Box as Box2,
  Text as Text2,
  Newline as InkNewline,
  Spacer as InkSpacer,
  useInput as useInput3
} from "ink";
import { marked } from "marked";

// src/hooks.ts
import { useState as useState4, useCallback as useCallback5, useRef as useRef3, useEffect as useEffect4 } from "react";
import {
  applySpecPatch,
  validateSpec,
  autoFixSpec,
  formatSpecIssues
} from "@json-render/core";
function useBoundProp(propValue, bindingPath) {
  const { set } = useStateStore();
  const setValue = useCallback5(
    (value) => {
      if (bindingPath) set(bindingPath, value);
    },
    [bindingPath, set]
  );
  return [propValue, setValue];
}
function looksLikeJson(line) {
  return line.startsWith("{") || line.startsWith("[");
}
function parsePatchLine(line) {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith("//")) {
    return { patch: null, malformed: false };
  }
  if (!looksLikeJson(trimmed)) {
    return { patch: null, malformed: false };
  }
  try {
    const parsed = JSON.parse(trimmed);
    if (parsed && typeof parsed === "object" && typeof parsed.op === "string") {
      return { patch: parsed, malformed: false };
    }
    return { patch: null, malformed: false };
  } catch {
  }
  let attempt = trimmed;
  for (let i = 0; i < 8; i++) {
    const last = attempt[attempt.length - 1];
    if (last === "}" || last === "]") {
      attempt = attempt.slice(0, -1);
      try {
        const result = JSON.parse(attempt);
        if (result && typeof result === "object" && typeof result.op === "string") {
          console.warn(
            `[json-render] Recovered malformed JSONL line by removing ${i + 1} trailing '${last}'`
          );
          return { patch: result, malformed: false };
        }
        return { patch: null, malformed: true };
      } catch {
      }
    } else {
      break;
    }
  }
  return { patch: null, malformed: true };
}
function useUIStream({
  api,
  onComplete,
  onError,
  fetch: fetchFn = globalThis.fetch,
  validate: enableValidation = false,
  maxRetries = 5
}) {
  const [spec, setSpec] = useState4(null);
  const [isStreaming, setIsStreaming] = useState4(false);
  const [error, setError] = useState4(null);
  const abortControllerRef = useRef3(null);
  const requestIdRef = useRef3(0);
  const onCompleteRef = useRef3(onComplete);
  onCompleteRef.current = onComplete;
  const onErrorRef = useRef3(onError);
  onErrorRef.current = onError;
  const stop = useCallback5(() => {
    abortControllerRef.current?.abort();
    setIsStreaming(false);
  }, []);
  const clear = useCallback5(() => {
    setSpec(null);
    setError(null);
  }, []);
  const streamRequest = useCallback5(
    async (prompt, context, initialSpec, abortOnMalformed, controller) => {
      let currentSpec = initialSpec;
      setSpec(currentSpec);
      const malformedLines = [];
      const response = await fetchFn(api, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          prompt,
          context,
          currentSpec
        }),
        signal: controller.signal
      });
      if (!response.ok) {
        let errorMessage = `HTTP error: ${response.status}`;
        try {
          const errorData = await response.json();
          if (errorData.message) {
            errorMessage = errorData.message;
          } else if (errorData.error) {
            errorMessage = errorData.error;
          }
        } catch {
        }
        throw new Error(errorMessage);
      }
      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error("No response body");
      }
      const decoder = new TextDecoder();
      let buffer = "";
      let aborted = false;
      while (true) {
        const { done, value } = await reader.read();
        if (done) {
          break;
        }
        const decoded = decoder.decode(value, { stream: true });
        buffer += decoded;
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";
        for (const line of lines) {
          const { patch, malformed } = parsePatchLine(line);
          if (patch) {
            currentSpec = applySpecPatch(structuredClone(currentSpec), patch);
            setSpec(currentSpec);
          } else if (malformed) {
            malformedLines.push(line.trim());
            if (abortOnMalformed) {
              await reader.cancel();
              aborted = true;
              break;
            }
          }
        }
        if (aborted) break;
      }
      if (!aborted && buffer.trim()) {
        const { patch, malformed } = parsePatchLine(buffer);
        if (patch) {
          currentSpec = applySpecPatch(structuredClone(currentSpec), patch);
          setSpec(currentSpec);
        } else if (malformed) {
          malformedLines.push(buffer.trim());
        }
      }
      return { spec: currentSpec, completed: !aborted, malformedLines };
    },
    [api, fetchFn]
  );
  const send = useCallback5(
    async (prompt, context) => {
      abortControllerRef.current?.abort();
      const controller = new AbortController();
      abortControllerRef.current = controller;
      const thisRequestId = ++requestIdRef.current;
      setIsStreaming(true);
      setError(null);
      const previousSpec = context?.previousSpec;
      let currentSpec = previousSpec && previousSpec.root ? { ...previousSpec, elements: { ...previousSpec.elements } } : { root: "", elements: {} };
      let retriesUsed = 0;
      let currentPrompt = prompt;
      let currentContext = context;
      try {
        while (retriesUsed <= maxRetries) {
          const result = await streamRequest(
            currentPrompt,
            currentContext,
            currentSpec,
            enableValidation,
            // only abort on malformed when validation is on
            controller
          );
          currentSpec = result.spec;
          if (!result.completed && result.malformedLines.length > 0) {
            if (retriesUsed >= maxRetries) {
              break;
            }
            retriesUsed++;
            currentContext = { ...context, previousSpec: currentSpec };
            currentPrompt = `The previous generation contained malformed JSON that could not be parsed. The line was:
${result.malformedLines[result.malformedLines.length - 1]?.slice(0, 500)}

The current spec state is provided. Continue generating from where you left off. Output ONLY the remaining patches needed to complete the UI.`;
            continue;
          }
          if (!enableValidation || !currentSpec.root) {
            break;
          }
          const { spec: fixedSpec, fixes } = autoFixSpec(currentSpec);
          if (fixes.length > 0) {
            currentSpec = fixedSpec;
            setSpec({ ...currentSpec });
          }
          const validation = validateSpec(currentSpec);
          if (validation.valid) {
            break;
          }
          if (retriesUsed >= maxRetries) {
            break;
          }
          retriesUsed++;
          const issueText = formatSpecIssues(validation.issues);
          currentContext = { ...context, previousSpec: currentSpec };
          currentPrompt = `FIX THE FOLLOWING ERRORS in the current UI spec. Output ONLY the patches needed to fix these issues, do not recreate the entire UI.

` + issueText;
        }
        if (enableValidation && retriesUsed >= maxRetries && currentSpec.root) {
          const finalValidation = validateSpec(currentSpec);
          if (!finalValidation.valid) {
            const issueText = formatSpecIssues(finalValidation.issues);
            const validationError = new Error(
              `Spec validation failed after ${maxRetries} retries:
${issueText}`
            );
            setError(validationError);
            onErrorRef.current?.(validationError);
            return;
          }
        }
        onCompleteRef.current?.(currentSpec);
      } catch (err) {
        if (err.name === "AbortError") {
          return;
        }
        const error2 = err instanceof Error ? err : new Error(String(err));
        setError(error2);
        onErrorRef.current?.(error2);
      } finally {
        if (requestIdRef.current === thisRequestId) {
          setIsStreaming(false);
        }
      }
    },
    [api, fetchFn, enableValidation, maxRetries, streamRequest]
  );
  useEffect4(() => {
    return () => {
      abortControllerRef.current?.abort();
    };
  }, []);
  return {
    spec,
    isStreaming,
    error,
    send,
    stop,
    clear
  };
}
function flatToTree(elements) {
  const elementMap = {};
  let root = "";
  for (const element of elements) {
    elementMap[element.key] = {
      type: element.type,
      props: element.props,
      children: [],
      visible: element.visible
    };
  }
  for (const element of elements) {
    if (element.parentKey) {
      const parent = elementMap[element.parentKey];
      if (parent) {
        if (!parent.children) {
          parent.children = [];
        }
        parent.children.push(element.key);
      } else {
        console.warn(
          `[json-render] flatToTree: element "${element.key}" references parent "${element.parentKey}" which does not exist. Element will be orphaned.`
        );
      }
    } else {
      if (root) {
        console.warn(
          `[json-render] flatToTree: multiple root elements found ("${root}" and "${element.key}"). Using "${element.key}" as root.`
        );
      }
      root = element.key;
    }
  }
  return { root, elements: elementMap };
}

// src/components/standard.tsx
import { Fragment, jsx as jsx7, jsxs as jsxs2 } from "react/jsx-runtime";
var BLOCKED_PROPS = /* @__PURE__ */ new Set([
  "key",
  "ref",
  "children",
  "dangerouslySetInnerHTML",
  "style",
  "className",
  "id"
]);
var INVISIBLE_COLORS = /* @__PURE__ */ new Set(["black", "#000", "#000000"]);
function safeColor(color) {
  if (color && INVISIBLE_COLORS.has(color)) return void 0;
  return color;
}
function safeBoxProps(obj) {
  const result = {};
  for (const [k, v] of Object.entries(obj)) {
    if (v !== void 0 && v !== null && !BLOCKED_PROPS.has(k)) {
      if (k === "color" && typeof v === "string" && INVISIBLE_COLORS.has(v)) {
        continue;
      }
      result[k] = v;
    }
  }
  return result;
}
function BoxComponent({ element, children }) {
  const p = element.props;
  return /* @__PURE__ */ jsx7(Box2, { ...safeBoxProps(p), children });
}
function TextComponent({ element }) {
  const { text, ...style } = element.props;
  return /* @__PURE__ */ jsx7(Text2, { ...safeBoxProps(style), children: text ?? "" });
}
function NewlineComponent({ element }) {
  const p = element.props;
  return /* @__PURE__ */ jsx7(InkNewline, { count: p.count });
}
function SpacerComponent() {
  return /* @__PURE__ */ jsx7(InkSpacer, {});
}
function HeadingComponent({ element }) {
  const p = element.props;
  const level = p.level ?? "h2";
  switch (level) {
    case "h1":
      return /* @__PURE__ */ jsx7(Text2, { bold: true, underline: true, color: safeColor(p.color), children: p.text ?? "" });
    case "h2":
      return /* @__PURE__ */ jsx7(Text2, { bold: true, color: safeColor(p.color), children: p.text ?? "" });
    case "h3":
      return /* @__PURE__ */ jsx7(Text2, { bold: true, dimColor: true, color: safeColor(p.color), children: p.text ?? "" });
    case "h4":
    default:
      return /* @__PURE__ */ jsx7(Text2, { dimColor: true, color: safeColor(p.color), children: p.text ?? "" });
  }
}
function DividerComponent({ element }) {
  const p = element.props;
  const char = Array.from(p.character ?? "\u2500")[0] || "\u2500";
  const totalWidth = p.width ?? 40;
  if (p.title) {
    const titleStr = ` ${p.title} `;
    const available = Math.max(0, totalWidth - titleStr.length);
    const leftLen = Math.floor(available / 2);
    const rightLen = available - leftLen;
    return /* @__PURE__ */ jsxs2(Text2, { color: safeColor(p.color), dimColor: p.dimColor, children: [
      char.repeat(leftLen),
      titleStr,
      char.repeat(rightLen)
    ] });
  }
  return /* @__PURE__ */ jsx7(Text2, { color: safeColor(p.color), dimColor: p.dimColor, children: char.repeat(totalWidth) });
}
var BADGE_COLORS = {
  default: "white",
  info: "blue",
  success: "green",
  warning: "yellow",
  error: "red"
};
function BadgeComponent({ element }) {
  const p = element.props;
  const color = BADGE_COLORS[p.variant ?? "default"] ?? "white";
  return /* @__PURE__ */ jsx7(Text2, { backgroundColor: color, bold: true, children: ` ${p.label ?? ""} ` });
}
var SPINNER_FRAMES = ["\u280B", "\u2819", "\u2839", "\u2838", "\u283C", "\u2834", "\u2826", "\u2827", "\u2807", "\u280F"];
function SpinnerComponent({ element }) {
  const p = element.props;
  const [frame, setFrame] = useState5(0);
  useEffect5(() => {
    const timer = setInterval(() => {
      setFrame((prev) => (prev + 1) % SPINNER_FRAMES.length);
    }, 80);
    return () => clearInterval(timer);
  }, []);
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    /* @__PURE__ */ jsx7(Text2, { color: safeColor(p.color), children: SPINNER_FRAMES[frame] }),
    p.label ? /* @__PURE__ */ jsx7(Text2, { children: p.label }) : null
  ] });
}
function ProgressBarComponent({ element }) {
  const p = element.props;
  const progress = Math.min(1, Math.max(0, p.progress ?? 0));
  const barWidth = p.width ?? 30;
  const filled = Math.round(progress * barWidth);
  const empty = barWidth - filled;
  const percentage = Math.round(progress * 100);
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    p.label ? /* @__PURE__ */ jsx7(Text2, { children: p.label }) : null,
    /* @__PURE__ */ jsxs2(Text2, { children: [
      /* @__PURE__ */ jsx7(Text2, { color: safeColor(p.color) ?? "green", children: "\u2588".repeat(filled) }),
      /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: "\u2591".repeat(empty) })
    ] }),
    /* @__PURE__ */ jsxs2(Text2, { dimColor: true, children: [
      percentage,
      "%"
    ] })
  ] });
}
var SPARK_BLOCKS = ["\u2581", "\u2582", "\u2583", "\u2584", "\u2585", "\u2586", "\u2587", "\u2588"];
function SparklineComponent({ element }) {
  const p = element.props;
  const data = p.data ?? [];
  if (data.length === 0) {
    return p.label ? /* @__PURE__ */ jsxs2(Text2, { dimColor: true, children: [
      p.label,
      ": (no data)"
    ] }) : null;
  }
  const minVal = p.min ?? Math.min(...data);
  const maxVal = p.max ?? Math.max(...data);
  const range = maxVal - minVal || 1;
  const maxWidth = p.width ?? data.length;
  let sampled = data;
  if (maxWidth < data.length) {
    sampled = [];
    for (let i = 0; i < maxWidth; i++) {
      const idx = maxWidth === 1 ? 0 : Math.round(i / (maxWidth - 1) * (data.length - 1));
      sampled.push(data[idx]);
    }
  }
  const blocks = sampled.map((v) => {
    const normalized = (v - minVal) / range;
    const idx = Math.min(
      SPARK_BLOCKS.length - 1,
      Math.round(normalized * (SPARK_BLOCKS.length - 1))
    );
    return SPARK_BLOCKS[idx];
  }).join("");
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    p.label ? /* @__PURE__ */ jsx7(Text2, { children: p.label }) : null,
    /* @__PURE__ */ jsx7(Text2, { color: safeColor(p.color) ?? "green", children: blocks })
  ] });
}
function BarChartComponent({ element }) {
  const p = element.props;
  const data = p.data ?? [];
  if (data.length === 0) return null;
  const maxValue = Math.max(...data.map((d) => d.value), 1);
  const total = data.reduce((sum, d) => sum + d.value, 0);
  const maxLabelLen = Math.max(...data.map((d) => d.label.length));
  const barWidth = p.width ?? 30;
  return /* @__PURE__ */ jsx7(Box2, { flexDirection: "column", children: data.map((item) => {
    const filled = Math.round(item.value / maxValue * barWidth);
    const padded = item.label.padEnd(maxLabelLen);
    const suffix = [];
    if (p.showValues) suffix.push(String(item.value));
    if (p.showPercentage && total > 0)
      suffix.push(`${Math.round(item.value / total * 100)}%`);
    return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
      /* @__PURE__ */ jsx7(Text2, { children: padded }),
      /* @__PURE__ */ jsx7(Text2, { color: item.color ?? "green", children: "\u2588".repeat(filled) }),
      suffix.length > 0 ? /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: suffix.join(" ") }) : null
    ] }, item.label);
  }) });
}
function TableComponent({ element }) {
  const p = element.props;
  const columns = p.columns ?? [];
  const rows = p.rows ?? [];
  const colWidths = columns.map((col) => {
    if (col.width) return col.width;
    let max = col.header.length;
    for (const row of rows) {
      const val = row[col.key] ?? "";
      if (val.length > max) max = val.length;
    }
    return max + 2;
  });
  const padCell = (text, width, align) => {
    const truncated = text.length > width ? text.slice(0, width - 1) + "\u2026" : text;
    const pad = width - truncated.length;
    if (align === "right") return " ".repeat(pad) + truncated;
    if (align === "center") {
      const left = Math.floor(pad / 2);
      return " ".repeat(left) + truncated + " ".repeat(pad - left);
    }
    return truncated + " ".repeat(pad);
  };
  const borderStyleProp = p.borderStyle;
  return /* @__PURE__ */ jsxs2(
    Box2,
    {
      flexDirection: "column",
      alignSelf: "flex-start",
      borderStyle: borderStyleProp,
      backgroundColor: p.backgroundColor,
      children: [
        /* @__PURE__ */ jsx7(Box2, { children: columns.map((col, i) => /* @__PURE__ */ jsx7(Text2, { bold: true, color: safeColor(p.headerColor), children: padCell(col.header, colWidths[i], col.align) }, col.key)) }),
        /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: colWidths.map((w) => "\u2500".repeat(w)).join("\u2500") }),
        rows.map((row, rowIdx) => {
          const rowKey = columns.map((c) => row[c.key] ?? "").join("\0") || String(rowIdx);
          return /* @__PURE__ */ jsx7(Box2, { gap: 0, children: columns.map((col, i) => /* @__PURE__ */ jsx7(Text2, { children: padCell(row[col.key] || "\u2014", colWidths[i], col.align) }, col.key)) }, rowKey);
        })
      ]
    }
  );
}
function ListComponent({ element }) {
  const p = element.props;
  const items = p.items ?? [];
  const bullet = p.bulletChar ?? "\u2022";
  return /* @__PURE__ */ jsx7(Box2, { flexDirection: "column", gap: p.spacing ?? 0, children: items.map((item, i) => /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: p.ordered ? `${i + 1}.` : bullet }),
    /* @__PURE__ */ jsx7(Text2, { children: item })
  ] }, `${i}:${item}`)) });
}
function ListItemComponent({ element }) {
  const p = element.props;
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    p.leading ? /* @__PURE__ */ jsx7(Text2, { children: p.leading }) : null,
    /* @__PURE__ */ jsxs2(Box2, { flexDirection: "column", flexGrow: 1, children: [
      /* @__PURE__ */ jsx7(Text2, { bold: true, children: p.title ?? "" }),
      p.subtitle ? /* @__PURE__ */ jsx7(Text2, { children: p.subtitle }) : null
    ] }),
    p.trailing ? /* @__PURE__ */ jsx7(Text2, { children: String(p.trailing) }) : null
  ] });
}
function CardComponent({ element, children }) {
  const p = element.props;
  return /* @__PURE__ */ jsxs2(
    Box2,
    {
      flexDirection: "column",
      alignSelf: "flex-start",
      backgroundColor: p.backgroundColor ?? "#1a1a1a",
      padding: p.padding ?? 1,
      children: [
        p.title ? /* @__PURE__ */ jsx7(Box2, { marginBottom: 1, children: /* @__PURE__ */ jsx7(Text2, { bold: true, children: p.title }) }) : null,
        children
      ]
    }
  );
}
function coerceToString(val) {
  if (val == null) return "";
  if (typeof val === "string") return val;
  if (typeof val === "number") return val.toLocaleString();
  if (typeof val === "boolean") return val ? "true" : "false";
  if (Array.isArray(val)) return val.map(coerceToString).join(", ");
  if (typeof val === "object") {
    const values = Object.values(val);
    return values.map(coerceToString).join(" ");
  }
  return String(val);
}
function KeyValueComponent({ element }) {
  const p = element.props;
  const sep = p.separator ?? ":";
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    /* @__PURE__ */ jsxs2(Text2, { color: safeColor(p.labelColor), bold: true, dimColor: true, children: [
      p.label ?? "",
      sep
    ] }),
    /* @__PURE__ */ jsx7(Text2, { children: coerceToString(p.value) || "\u2014" })
  ] });
}
function LinkComponent({ element }) {
  const p = element.props;
  const url = p.url ?? "";
  const label = p.label ?? url;
  return /* @__PURE__ */ jsxs2(Text2, { color: safeColor(p.color) ?? "cyan", underline: true, children: [
    label,
    label !== url ? ` (${url})` : ""
  ] });
}
var STATUS_CONFIG = {
  info: { icon: "\u2139", color: "blue" },
  success: { icon: "\u2714", color: "green" },
  warning: { icon: "\u26A0", color: "yellow" },
  error: { icon: "\u2716", color: "red" }
};
function StatusLineComponent({ element }) {
  const p = element.props;
  const status = p.status ?? "info";
  const config = STATUS_CONFIG[status];
  const icon = p.icon ?? config?.icon ?? "\u2022";
  const color = config?.color ?? "white";
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    /* @__PURE__ */ jsx7(Text2, { color, children: icon }),
    /* @__PURE__ */ jsx7(Text2, { children: p.text ?? "" })
  ] });
}
var TREND_CONFIG = {
  up: { prefix: "+", color: "green" },
  down: { prefix: "", color: "red" },
  neutral: { prefix: "~", color: "gray" }
};
function MetricComponent({ element }) {
  const p = element.props;
  const trend = p.trend ? TREND_CONFIG[p.trend] : null;
  return /* @__PURE__ */ jsxs2(Box2, { flexDirection: "column", children: [
    /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: p.label ?? "" }),
    /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
      /* @__PURE__ */ jsx7(Text2, { bold: true, children: p.value ?? "" }),
      trend && p.detail ? /* @__PURE__ */ jsxs2(Text2, { color: trend.color, children: [
        trend.prefix,
        p.detail
      ] }) : null
    ] }),
    !trend && p.detail ? /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: p.detail }) : null
  ] });
}
var CALLOUT_CONFIG = {
  info: { label: "INFO", color: "blue" },
  tip: { label: "TIP", color: "green" },
  warning: { label: "WARNING", color: "yellow" },
  important: { label: "IMPORTANT", color: "magenta" }
};
function CalloutComponent({ element }) {
  const p = element.props;
  const config = CALLOUT_CONFIG[p.type ?? "info"] ?? CALLOUT_CONFIG.info;
  return /* @__PURE__ */ jsxs2(
    Box2,
    {
      flexDirection: "column",
      borderStyle: "bold",
      borderLeft: true,
      borderRight: false,
      borderTop: false,
      borderBottom: false,
      borderColor: config.color,
      paddingLeft: 1,
      children: [
        p.title ? /* @__PURE__ */ jsx7(Text2, { bold: true, color: config.color, children: p.title }) : null,
        /* @__PURE__ */ jsx7(Text2, { children: p.content ?? "" })
      ]
    }
  );
}
var TIMELINE_DOT = {
  completed: { dot: "\u25CF", color: "green" },
  current: { dot: "\u25C6", color: "cyan" },
  upcoming: { dot: "\u25CB", color: "gray" }
};
function TimelineComponent({ element }) {
  const p = element.props;
  const items = p.items ?? [];
  return /* @__PURE__ */ jsx7(Box2, { flexDirection: "column", gap: 1, children: items.map((item, i) => {
    const cfg = TIMELINE_DOT[item.status ?? "upcoming"] ?? TIMELINE_DOT.upcoming;
    return /* @__PURE__ */ jsxs2(Box2, { flexDirection: "column", children: [
      /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
        /* @__PURE__ */ jsx7(Text2, { color: cfg.color, children: cfg.dot }),
        /* @__PURE__ */ jsx7(Text2, { bold: true, children: item.title ?? "" }),
        item.date ? /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: item.date }) : null
      ] }),
      item.description ? /* @__PURE__ */ jsx7(Box2, { paddingLeft: 2, children: /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: item.description }) }) : null
    ] }, i);
  }) });
}
function codePointLength(s) {
  return Array.from(s).length;
}
function codePointSlice(s, start, end) {
  return Array.from(s).slice(start, end).join("");
}
function TextInputComponent({ element, emit, bindings }) {
  const p = element.props;
  const { isActive } = useFocus();
  const [value, setValue] = useBoundProp(
    p.value ?? "",
    bindings?.value
  );
  const currentValue = value ?? "";
  const [cursorPos, setCursorPos] = useState5(codePointLength(currentValue));
  useEffect5(() => {
    setCursorPos((prev) => Math.min(prev, codePointLength(currentValue)));
  }, [currentValue]);
  useInput3(
    (input, key) => {
      if (key.return) {
        emit("submit");
        return;
      }
      if (key.backspace || key.delete) {
        if (cursorPos > 0) {
          const newVal = codePointSlice(currentValue, 0, cursorPos - 1) + codePointSlice(currentValue, cursorPos);
          setValue(newVal);
          setCursorPos((prev) => prev - 1);
          emit("change");
        }
        return;
      }
      if (key.leftArrow) {
        setCursorPos((prev) => Math.max(0, prev - 1));
        return;
      }
      if (key.rightArrow) {
        setCursorPos(
          (prev) => Math.min(codePointLength(currentValue), prev + 1)
        );
        return;
      }
      if (key.ctrl || key.meta || key.escape || key.tab) return;
      if (key.upArrow || key.downArrow) return;
      if (input) {
        const newVal = codePointSlice(currentValue, 0, cursorPos) + input + codePointSlice(currentValue, cursorPos);
        setValue(newVal);
        setCursorPos((prev) => prev + codePointLength(input));
        emit("change");
      }
    },
    { isActive }
  );
  const displayValue = currentValue ? p.mask ? p.mask.repeat(codePointLength(currentValue)) : currentValue : "";
  const before = codePointSlice(displayValue, 0, cursorPos);
  const after = codePointSlice(displayValue, cursorPos);
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    p.label ? /* @__PURE__ */ jsxs2(Text2, { bold: true, children: [
      p.label,
      ":"
    ] }) : null,
    /* @__PURE__ */ jsx7(Text2, { children: currentValue ? /* @__PURE__ */ jsxs2(Fragment, { children: [
      before,
      isActive ? /* @__PURE__ */ jsx7(Text2, { color: "cyan", children: "\u258E" }) : null,
      after
    ] }) : /* @__PURE__ */ jsxs2(Fragment, { children: [
      isActive ? /* @__PURE__ */ jsx7(Text2, { color: "cyan", children: "\u258E" }) : null,
      /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: p.placeholder ?? "" })
    ] }) })
  ] });
}
function SelectComponent({ element, emit, bindings }) {
  const p = element.props;
  const { isActive } = useFocus();
  const options = p.options ?? [];
  const [selectedValue, setSelectedValue] = useBoundProp(
    p.value ?? "",
    bindings?.value
  );
  const currentIndex = options.findIndex((o) => o.value === selectedValue);
  const [highlightIndex, setHighlightIndex] = useState5(
    currentIndex >= 0 ? currentIndex : 0
  );
  const optionsKey = options.map((o) => o.value).join("\0");
  useEffect5(() => {
    const idx = options.findIndex((o) => o.value === selectedValue);
    if (idx >= 0) {
      setHighlightIndex(idx);
    } else if (options.length > 0) {
      setHighlightIndex((prev) => Math.min(prev, options.length - 1));
    } else {
      setHighlightIndex(0);
    }
  }, [selectedValue, optionsKey]);
  useInput3(
    (_input, key) => {
      if (options.length === 0) return;
      if (key.upArrow) {
        setHighlightIndex((prev) => prev > 0 ? prev - 1 : options.length - 1);
      } else if (key.downArrow) {
        setHighlightIndex((prev) => prev < options.length - 1 ? prev + 1 : 0);
      } else if (key.return) {
        const option = options[highlightIndex];
        if (option) {
          setSelectedValue(option.value);
          emit("change");
        }
      }
    },
    { isActive }
  );
  return /* @__PURE__ */ jsxs2(Box2, { flexDirection: "column", children: [
    p.label ? /* @__PURE__ */ jsx7(Box2, { marginBottom: 1, children: /* @__PURE__ */ jsxs2(Text2, { bold: true, children: [
      p.label,
      ":"
    ] }) }) : null,
    options.map((option, i) => {
      const isHighlighted = i === highlightIndex;
      const isSelected = option.value === selectedValue;
      return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
        /* @__PURE__ */ jsx7(Text2, { color: isHighlighted ? "cyan" : void 0, children: isHighlighted ? "\u276F" : " " }),
        /* @__PURE__ */ jsx7(
          Text2,
          {
            color: isSelected ? "green" : isHighlighted ? "cyan" : void 0,
            bold: isSelected,
            children: option.label
          }
        )
      ] }, option.value);
    })
  ] });
}
function MultiSelectComponent({
  element,
  emit,
  bindings
}) {
  const p = element.props;
  const { isActive } = useFocus();
  const options = p.options ?? [];
  const [selectedValues, setSelectedValues] = useBoundProp(
    p.value ?? [],
    bindings?.value
  );
  const selected = selectedValues ?? [];
  const [highlightIndex, setHighlightIndex] = useState5(0);
  const optionsKey = options.map((o) => o.value).join("\0");
  useEffect5(() => {
    if (options.length > 0) {
      setHighlightIndex((prev) => Math.min(prev, options.length - 1));
    } else {
      setHighlightIndex(0);
    }
  }, [optionsKey]);
  useInput3(
    (_input, key) => {
      if (options.length === 0) return;
      if (key.upArrow) {
        setHighlightIndex((prev) => prev > 0 ? prev - 1 : options.length - 1);
      } else if (key.downArrow) {
        setHighlightIndex((prev) => prev < options.length - 1 ? prev + 1 : 0);
      } else if (_input === " ") {
        const option = options[highlightIndex];
        if (!option) return;
        const isSelected = selected.includes(option.value);
        let next;
        if (isSelected) {
          if (p.min != null && selected.length <= p.min) return;
          next = selected.filter((v) => v !== option.value);
        } else {
          if (p.max != null && selected.length >= p.max) return;
          next = [...selected, option.value];
        }
        setSelectedValues(next);
        emit("change");
      } else if (key.return) {
        emit("submit");
      }
    },
    { isActive }
  );
  return /* @__PURE__ */ jsxs2(Box2, { flexDirection: "column", children: [
    p.label ? /* @__PURE__ */ jsxs2(Box2, { marginBottom: 1, children: [
      /* @__PURE__ */ jsxs2(Text2, { bold: true, children: [
        p.label,
        ":"
      ] }),
      /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: " (space to toggle, enter to confirm)" })
    ] }) : null,
    options.map((option, i) => {
      const isHighlighted = i === highlightIndex;
      const isChecked = selected.includes(option.value);
      return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
        /* @__PURE__ */ jsx7(Text2, { color: isHighlighted ? "cyan" : void 0, children: isHighlighted ? "\u276F" : " " }),
        /* @__PURE__ */ jsx7(Text2, { color: isChecked ? "green" : void 0, children: isChecked ? "\u25C9" : "\u25EF" }),
        /* @__PURE__ */ jsx7(Text2, { color: isHighlighted ? "cyan" : void 0, bold: isChecked, children: option.label })
      ] }, option.value);
    }),
    selected.length > 0 ? /* @__PURE__ */ jsx7(Box2, { marginTop: 1, children: /* @__PURE__ */ jsxs2(Text2, { dimColor: true, children: [
      "Selected: ",
      selected.length
    ] }) }) : null
  ] });
}
function ConfirmInputComponent({ element, emit }) {
  const p = element.props;
  const { isActive } = useFocus();
  const [value, setValue] = useState5(p.defaultValue ?? null);
  useInput3(
    (input, key) => {
      const lower = input.toLowerCase();
      if (lower === "y") {
        setValue(true);
        emit("confirm");
      } else if (lower === "n") {
        setValue(false);
        emit("deny");
      } else if (key.return && value !== null) {
        emit(value ? "confirm" : "deny");
      }
    },
    { isActive }
  );
  const yesLabel = p.yesLabel ?? "Yes";
  const noLabel = p.noLabel ?? "No";
  return /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
    /* @__PURE__ */ jsx7(Text2, { bold: true, children: p.message ?? "Are you sure?" }),
    /* @__PURE__ */ jsxs2(Text2, { children: [
      /* @__PURE__ */ jsx7(Text2, { color: value === true ? "green" : "gray", bold: value === true, children: yesLabel }),
      /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: " / " }),
      /* @__PURE__ */ jsx7(Text2, { color: value === false ? "red" : "gray", bold: value === false, children: noLabel })
    ] }),
    /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: "(y/n)" })
  ] });
}
function TabsComponent({
  element,
  emit,
  bindings,
  children
}) {
  const p = element.props;
  const { isActive } = useFocus();
  const tabs = p.tabs ?? [];
  const [activeTab, setActiveTab] = useBoundProp(
    p.value ?? tabs[0]?.value ?? "",
    bindings?.value
  );
  const activeIndex = tabs.findIndex((t) => t.value === activeTab);
  useInput3(
    (_input, key) => {
      if (tabs.length === 0) return;
      const currentIdx = activeIndex >= 0 ? activeIndex : 0;
      if (key.leftArrow) {
        const next = currentIdx > 0 ? currentIdx - 1 : tabs.length - 1;
        const tab = tabs[next];
        if (tab) {
          setActiveTab(tab.value);
          emit("change");
        }
      } else if (key.rightArrow) {
        const next = currentIdx < tabs.length - 1 ? currentIdx + 1 : 0;
        const tab = tabs[next];
        if (tab) {
          setActiveTab(tab.value);
          emit("change");
        }
      }
    },
    { isActive }
  );
  const accentColor = p.color ?? "cyan";
  return /* @__PURE__ */ jsxs2(Box2, { flexDirection: "column", children: [
    /* @__PURE__ */ jsx7(Box2, { gap: 2, children: tabs.map((tab) => {
      const isActive2 = tab.value === activeTab;
      return /* @__PURE__ */ jsxs2(
        Text2,
        {
          color: isActive2 ? accentColor : void 0,
          bold: isActive2,
          dimColor: !isActive2,
          children: [
            tab.icon ? `${tab.icon} ` : "",
            tab.label
          ]
        },
        tab.value
      );
    }) }),
    /* @__PURE__ */ jsx7(Text2, { color: accentColor, children: tabs.map((tab) => {
      const isActive2 = tab.value === activeTab;
      const labelWidth = (tab.icon ? tab.icon.length + 1 : 0) + tab.label.length;
      return isActive2 ? "\u2501".repeat(labelWidth) + "  " : " ".repeat(labelWidth) + "  ";
    }).join("") }),
    children
  ] });
}
function renderInline(tokens) {
  const nodes = [];
  for (let i = 0; i < tokens.length; i++) {
    const t = tokens[i];
    switch (t.type) {
      case "strong":
        nodes.push(
          /* @__PURE__ */ jsx7(Text2, { bold: true, children: renderInline(t.tokens) }, i)
        );
        break;
      case "em":
        nodes.push(
          /* @__PURE__ */ jsx7(Text2, { italic: true, children: renderInline(t.tokens) }, i)
        );
        break;
      case "codespan":
        nodes.push(
          /* @__PURE__ */ jsx7(
            Text2,
            {
              bold: true,
              dimColor: true,
              children: `\`${t.text}\``
            },
            i
          )
        );
        break;
      case "del":
        nodes.push(
          /* @__PURE__ */ jsx7(Text2, { strikethrough: true, children: renderInline(t.tokens) }, i)
        );
        break;
      case "link":
        nodes.push(
          /* @__PURE__ */ jsx7(Text2, { underline: true, children: renderInline(t.tokens) }, i)
        );
        break;
      case "text": {
        const tt = t;
        if (tt.tokens) {
          nodes.push(...renderInline(tt.tokens));
        } else {
          nodes.push(tt.text);
        }
        break;
      }
      case "br":
        nodes.push("\n");
        break;
      case "escape":
        nodes.push(t.text);
        break;
      default:
        if ("text" in t) nodes.push(t.text);
        break;
    }
  }
  return nodes;
}
function renderBlock(token, key) {
  switch (token.type) {
    case "heading": {
      const h = token;
      const content = renderInline(h.tokens);
      if (h.depth === 1)
        return /* @__PURE__ */ jsx7(Text2, { bold: true, underline: true, children: content }, key);
      if (h.depth === 2)
        return /* @__PURE__ */ jsx7(Text2, { bold: true, children: content }, key);
      if (h.depth === 3)
        return /* @__PURE__ */ jsx7(Text2, { bold: true, dimColor: true, children: content }, key);
      return /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: content }, key);
    }
    case "paragraph": {
      const p = token;
      return /* @__PURE__ */ jsx7(Text2, { children: renderInline(p.tokens) }, key);
    }
    case "code": {
      const c = token;
      return /* @__PURE__ */ jsx7(
        Box2,
        {
          borderStyle: "single",
          borderColor: "gray",
          paddingX: 1,
          marginY: 0,
          children: /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: c.text })
        },
        key
      );
    }
    case "blockquote": {
      const bq = token;
      return /* @__PURE__ */ jsx7(
        Box2,
        {
          paddingLeft: 1,
          borderStyle: "bold",
          borderLeft: true,
          borderTop: false,
          borderBottom: false,
          borderRight: false,
          borderColor: "gray",
          children: /* @__PURE__ */ jsx7(Box2, { flexDirection: "column", children: bq.tokens.map((child, i) => renderBlock(child, i)) })
        },
        key
      );
    }
    case "list": {
      const l = token;
      return /* @__PURE__ */ jsx7(Box2, { flexDirection: "column", children: l.items.map((item, idx) => /* @__PURE__ */ jsxs2(Box2, { gap: 1, children: [
        /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: l.ordered ? `${Number(l.start ?? 1) + idx}.` : "\u2022" }),
        /* @__PURE__ */ jsx7(Text2, { children: item.tokens.flatMap((child) => {
          if (child.type === "text") {
            const tt = child;
            return tt.tokens ? renderInline(tt.tokens) : [tt.text];
          }
          if (child.type === "list") return [renderBlock(child, idx)];
          return "tokens" in child ? renderInline(child.tokens) : [];
        }) })
      ] }, idx)) }, key);
    }
    case "hr":
      return /* @__PURE__ */ jsx7(Text2, { dimColor: true, children: "\u2500".repeat(40) }, key);
    case "space":
      return null;
    default:
      if ("text" in token)
        return /* @__PURE__ */ jsx7(Text2, { children: token.text }, key);
      return null;
  }
}
function MarkdownText({ text }) {
  const tokens = useMemo7(
    () => marked.lexer(text, { gfm: true, breaks: true }),
    [text]
  );
  const blocks = tokens.map((t, i) => renderBlock(t, i)).filter((n) => n != null);
  return /* @__PURE__ */ jsx7(Box2, { flexDirection: "column", gap: 0, children: blocks });
}
function MarkdownComponent({ element }) {
  const p = element.props;
  return /* @__PURE__ */ jsx7(MarkdownText, { text: p.text ?? "" });
}
var standardComponents = {
  Box: BoxComponent,
  Text: TextComponent,
  Newline: NewlineComponent,
  Spacer: SpacerComponent,
  Heading: HeadingComponent,
  Divider: DividerComponent,
  Badge: BadgeComponent,
  Spinner: SpinnerComponent,
  ProgressBar: ProgressBarComponent,
  Sparkline: SparklineComponent,
  BarChart: BarChartComponent,
  Table: TableComponent,
  List: ListComponent,
  ListItem: ListItemComponent,
  Card: CardComponent,
  KeyValue: KeyValueComponent,
  Link: LinkComponent,
  StatusLine: StatusLineComponent,
  Metric: MetricComponent,
  Callout: CalloutComponent,
  Timeline: TimelineComponent,
  TextInput: TextInputComponent,
  Select: SelectComponent,
  MultiSelect: MultiSelectComponent,
  ConfirmInput: ConfirmInputComponent,
  Tabs: TabsComponent,
  Markdown: MarkdownComponent
};

// src/renderer.tsx
import { Fragment as Fragment2, jsx as jsx8, jsxs as jsxs3 } from "react/jsx-runtime";
var ElementErrorBoundary = class extends React3.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false, errorType: null };
  }
  static getDerivedStateFromError() {
    return { hasError: true, errorType: "__pending__" };
  }
  static getDerivedStateFromProps(props, state) {
    if (state.hasError && state.errorType !== null && state.errorType !== "__pending__" && state.errorType !== props.elementType) {
      return { hasError: false, errorType: null };
    }
    return null;
  }
  componentDidCatch(error, info) {
    this.setState({ errorType: this.props.elementType });
    console.error(
      `[json-render] Rendering error in <${this.props.elementType}>:`,
      error,
      info.componentStack
    );
  }
  render() {
    if (this.state.hasError) {
      return null;
    }
    return this.props.children;
  }
};
var ElementRenderer = React3.memo(function ElementRenderer2({
  element,
  spec,
  registry,
  loading,
  fallback
}) {
  const repeatScope = useRepeatScope();
  const { ctx } = useVisibility();
  const { execute } = useActions();
  const { getSnapshot } = useStateStore();
  const fullCtx = useMemo8(
    () => repeatScope ? {
      ...ctx,
      repeatItem: repeatScope.item,
      repeatIndex: repeatScope.index,
      repeatBasePath: repeatScope.basePath
    } : ctx,
    [ctx, repeatScope]
  );
  const isVisible = element.visible === void 0 ? true : evaluateVisibility2(element.visible, fullCtx);
  const onBindings = element.on;
  const emit = useCallback6(
    (eventName) => {
      const binding = onBindings?.[eventName];
      if (!binding) return;
      const actionBindings = Array.isArray(binding) ? binding : [binding];
      (async () => {
        for (const b of actionBindings) {
          if (!b.params) {
            await execute(b);
            continue;
          }
          const liveCtx = {
            ...fullCtx,
            stateModel: getSnapshot()
          };
          const resolved = {};
          for (const [key, val] of Object.entries(b.params)) {
            resolved[key] = resolveActionParam(val, liveCtx);
          }
          await execute({ ...b, params: resolved });
        }
      })().catch((err) => {
        console.error(
          `[json-render] Error handling event "${eventName}":`,
          err
        );
      });
    },
    [onBindings, execute, fullCtx, getSnapshot]
  );
  if (!isVisible) {
    return null;
  }
  const rawProps = element.props;
  const elementBindings = resolveBindings(rawProps, fullCtx);
  const resolvedProps = resolveElementProps(rawProps, fullCtx);
  const resolvedElement = resolvedProps !== element.props ? { ...element, props: resolvedProps } : element;
  const Component = registry[resolvedElement.type] ?? fallback;
  if (!Component) {
    console.warn(
      `[json-render] No renderer for component type: ${resolvedElement.type}`
    );
    return null;
  }
  const children = resolvedElement.repeat ? /* @__PURE__ */ jsx8(
    RepeatChildren,
    {
      element: resolvedElement,
      spec,
      registry,
      loading,
      fallback
    }
  ) : resolvedElement.children?.map((childKey) => {
    const childElement = spec.elements[childKey];
    if (!childElement) {
      if (!loading) {
        console.warn(
          `[json-render] Missing element "${childKey}" referenced as child of "${resolvedElement.type}". This element will not render.`
        );
        return /* @__PURE__ */ jsx8(Text3, { color: "red", children: `[Missing: ${childKey}]` }, childKey);
      }
      return null;
    }
    return /* @__PURE__ */ jsx8(
      ElementRenderer2,
      {
        element: childElement,
        spec,
        registry,
        loading,
        fallback
      },
      childKey
    );
  });
  return /* @__PURE__ */ jsx8(ElementErrorBoundary, { elementType: resolvedElement.type, children: /* @__PURE__ */ jsx8(
    Component,
    {
      element: resolvedElement,
      emit,
      bindings: elementBindings,
      loading,
      children
    }
  ) });
});
function RepeatChildren({
  element,
  spec,
  registry,
  loading,
  fallback
}) {
  const { state } = useStateStore();
  const repeat = element.repeat;
  const statePath = repeat.statePath;
  const raw = getByPath2(state, statePath);
  const items = Array.isArray(raw) ? raw : [];
  return /* @__PURE__ */ jsx8(Fragment2, { children: items.map((itemValue, index) => {
    const key = repeat.key && typeof itemValue === "object" && itemValue !== null ? String(
      itemValue[repeat.key] ?? index
    ) : String(index);
    return /* @__PURE__ */ jsx8(
      RepeatScopeProvider,
      {
        item: itemValue,
        index,
        basePath: `${statePath}/${index}`,
        children: element.children?.map((childKey) => {
          const childElement = spec.elements[childKey];
          if (!childElement) {
            if (!loading) {
              console.warn(
                `[json-render] Missing element "${childKey}" referenced as child of "${element.type}" (repeat). This element will not render.`
              );
              return /* @__PURE__ */ jsx8(
                Text3,
                {
                  color: "red",
                  children: `[Missing: ${childKey}]`
                },
                `${key}:${childKey}`
              );
            }
            return null;
          }
          return /* @__PURE__ */ jsx8(
            ElementRenderer,
            {
              element: childElement,
              spec,
              registry,
              loading,
              fallback
            },
            `${key}:${childKey}`
          );
        })
      },
      key
    );
  }) });
}
function Renderer({
  spec,
  registry: customRegistry,
  includeStandard = true,
  loading,
  fallback
}) {
  const registry = useMemo8(
    () => ({
      ...includeStandard ? standardComponents : {},
      ...customRegistry
    }),
    [customRegistry, includeStandard]
  );
  if (!spec || !spec.root) {
    return null;
  }
  const rootElement = spec.elements[spec.root];
  if (!rootElement) {
    return null;
  }
  return /* @__PURE__ */ jsx8(
    ElementRenderer,
    {
      element: rootElement,
      spec,
      registry,
      loading,
      fallback
    }
  );
}
function JSONUIProvider({
  store,
  initialState,
  handlers,
  navigate,
  validationFunctions,
  onStateChange,
  children
}) {
  return /* @__PURE__ */ jsx8(
    StateProvider,
    {
      store,
      initialState,
      onStateChange,
      children: /* @__PURE__ */ jsx8(VisibilityProvider, { children: /* @__PURE__ */ jsx8(ValidationProvider, { customFunctions: validationFunctions, children: /* @__PURE__ */ jsx8(ActionProvider, { handlers, navigate, children: /* @__PURE__ */ jsxs3(FocusProvider, { children: [
        children,
        /* @__PURE__ */ jsx8(ConfirmationDialogManager, {})
      ] }) }) }) })
    }
  );
}
function ConfirmationDialogManager() {
  const { pendingConfirmation, confirm, cancel } = useActions();
  if (!pendingConfirmation?.action.confirm) {
    return null;
  }
  return /* @__PURE__ */ jsx8(
    ConfirmDialog,
    {
      confirm: pendingConfirmation.action.confirm,
      onConfirm: confirm,
      onCancel: cancel
    }
  );
}
function defineRegistry(_catalog, options) {
  const registry = {};
  if (options.components) {
    for (const [name, componentFn] of Object.entries(options.components)) {
      registry[name] = ({
        element,
        children,
        emit,
        bindings,
        loading
      }) => {
        return componentFn({
          props: element.props,
          children,
          emit,
          bindings,
          loading
        });
      };
    }
  }
  const actionMap = options.actions ? Object.entries(options.actions) : [];
  const handlers = (getSetState, getState) => {
    const result = {};
    for (const [name, actionFn] of actionMap) {
      result[name] = async (params) => {
        const setState = getSetState();
        const state = getState();
        if (setState) {
          await actionFn(params, setState, state);
        } else {
          console.warn(
            `[json-render] Action "${name}" skipped: setState not available. Ensure the action handler is used within a mounted StateProvider.`
          );
        }
      };
    }
    return result;
  };
  const executeAction2 = async (actionName, params, setState, state = {}) => {
    const entry = actionMap.find(([name]) => name === actionName);
    if (entry) {
      await entry[1](params, setState, state);
    } else {
      console.warn(`[json-render] Unknown action: ${actionName}`);
    }
  };
  return { registry, handlers, executeAction: executeAction2 };
}
function createRenderer(_catalog, components) {
  const registry = components;
  return function CatalogRenderer({
    spec,
    store,
    state,
    onAction,
    onStateChange,
    loading,
    fallback
  }) {
    const actionHandlers = useMemo8(
      () => onAction ? new Proxy(
        {},
        {
          get: (_target, prop) => {
            return (params) => onAction(prop, params);
          },
          has: () => true
        }
      ) : void 0,
      [onAction]
    );
    return /* @__PURE__ */ jsx8(
      JSONUIProvider,
      {
        store,
        initialState: state,
        handlers: actionHandlers,
        onStateChange,
        children: /* @__PURE__ */ jsx8(
          Renderer,
          {
            spec,
            registry,
            loading,
            fallback
          }
        )
      }
    );
  };
}
export {
  ActionProvider,
  ConfirmDialog,
  FocusProvider,
  JSONUIProvider,
  Renderer,
  RepeatScopeProvider,
  StateProvider,
  ValidationProvider,
  VisibilityProvider,
  createRenderer,
  createStateStore2 as createStateStore,
  defineRegistry,
  flatToTree,
  schema,
  standardActionDefinitions,
  standardComponentDefinitions,
  standardComponents,
  useAction,
  useActions,
  useBoundProp,
  useFieldValidation,
  useFocus,
  useFocusDisable,
  useIsVisible,
  useOptionalValidation,
  useRepeatScope,
  useStateBinding,
  useStateStore,
  useStateValue,
  useUIStream,
  useValidation,
  useVisibility
};
